cmd/opm/alpha/diff/cmd.go

```diff
@@ -0,0 +1,157 @@
+package diff
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+	"k8s.io/kubectl/pkg/util/templates"
+
+	"github.com/operator-framework/operator-registry/internal/action"
+	"github.com/operator-framework/operator-registry/internal/declcfg"
+	containerd "github.com/operator-framework/operator-registry/pkg/image/containerdregistry"
+	"github.com/operator-framework/operator-registry/pkg/lib/certs"
+)
+
+const (
+	retryInterval = time.Second * 5
+	timeout       = time.Minute * 1
+)
+
+type diff struct {
+	oldRefs []string
+	newRefs []string
+
+	output string
+	caFile string
+
+	debug  bool
+	logger *logrus.Entry
+}
+
+func NewCmd() *cobra.Command {
+	a := diff{
+		logger: logrus.NewEntry(logrus.New()),
+	}
+	cmd := &cobra.Command{
+		Use:   "diff [--old refs]... --new refs...",
+		Short: "Diff old and new catalog references into a declarative config",
+		Long: templates.LongDesc(`
+Diff a set of old and new catalog references ("refs") to produce a
+declarative config containing only packages channels, and versions not present
+in the old set. This is known as "latest" mode.
+These references are passed through 'opm render' to produce a single declarative config.
+
+This command has special behavior when --old is omitted, called "heads-only" mode:
+instead of the output being that of 'opm render refs...'
+(which would be the case given the preceding behavior description),
+only the channel heads of all channels in all packages are included in the output,
+and dependencies. Dependencies are assumed to be provided by either an old ref,
+in which case they are not included in the diff, or a new ref, in which
+case they are included. Dependencies provided by some catalog unknown to
+'opm alpha diff' will not cause the command to error, but an error will occur
+if that catalog is not serving these dependencies at runtime.
+`),
+		Example: templates.Examples(`
+# Diff a catalog at some old state and latest state into a declarative config index.
+mkdir -p catalog-index
+opm alpha diff --old registry.org/my-catalog:abc123 --new registry.org/my-catalog:def456 -o yaml > my-catalog-index/index.yaml
+
+# Build and push this index into an index image.
+cat <<EOF > index.Dockerfile
+FROM quay.io/operator-framework/upstream-registry-builder:latest AS builder
+
+FROM scratch
+LABEL operators.operatorframework.io.index.configs.v1=/index/index.yaml
+COPY ./my-catalog-index /index
+COPY --from=builder /bin/opm /opm
+COPY --from=builder /bin/grpc_health_probe /bin/grpc_health_probe
+EXPOSE 50051
+ENTRYPOINT ["/opm"]
+CMD ["serve", "/index"]
+EOF
```

Comment:
This will likely need to change as soon as @ankitathomas's upcoming PR to build DC index images merges.

Reply:
Sounds good.

Reply:
Updated.